src/OpenDebugAD7/AD7DebugSession.cs

```diff
@@ -2007,6 +2008,46 @@ protected override void HandleSetFunctionBreakpointsRequestAsync(IRequestRespond
             responder.SetResponse(response);
         }
 
+        protected override void HandleCompletionsRequestAsync(IRequestResponder<CompletionsArguments, CompletionsResponse> responder)
+        {
+            if (!m_isStopped)
+            {
+                responder.SetError(new ProtocolException("Failed to handle CompletionsRequest", new Message(1105, AD7Resources.Error_TargetNotStopped)));
+                return;
+            }
+
+            IDebugStackFrame2 frame = null;
+            int? frameId = responder.Arguments.FrameId;
+            bool success = false;
+            if (frameId != null)
+                success = m_frameHandles.TryGet(frameId.Value, out frame);
+
+            try
+            {
+                MethodInfo methodInfo = m_program.GetType().GetMethod("AutoComplete");
```

Comment:
Rather than invoking through reflection, we should add a new interface for this to src\Microsoft.VisualStudio.Debugger.Interop.DAP\Microsoft.VisualStudio.Debugger.Interop.DAP.cs

Reply:
Please provide the interface.

Reply:
I think you can extend `IDebugProgramDAP`.

https://github.com/microsoft/MIEngine/blob/728f403c563c2c50c46542e9b44271a7f8d5895b/src/Microsoft.VisualStudio.Debugger.Interop.DAP/Microsoft.VisualStudio.Debugger.Interop.DAP.cs#L46

You will create `IDebugProgramDAP.AutoComplete(string command, IDebugStackFrame2 stackFrame, out string[] result)`.

Then in AD7Engine, you will implement `AutoComplete(string command, IDebugStackFrame2 stackFrame, out string[] result)` and it can call `_debuggedProcess.MICommandFactory.AutoComplete(..)`

https://github.com/microsoft/MIEngine/blob/73db59d89b8f0fbfdaa3cee1423e674e092ffc0f/src/MIDebugEngine/AD7.Impl/AD7Engine.cs#L38


You can use `IDebugProgramDAP.GetPointerSize` as an example.

@gregg-miskelly Is this correct or did you have another idea?

Reply:
Yup, that is my suggestion.

Reply:
Though if your goal is just to auto complete on commands, probably we should rename this to `AutoCompleteCommand` so that we could later have an `AutoCompleteExpression` if we ever wanted to support that.